File: android/app/src/main/java/dev/panelflow/WorkerHost.kt
package dev.panelflow

import android.annotation.SuppressLint
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.webkit.WebViewAssetLoader
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * The offscreen WebView that runs the shared store core, and the switchboard in
 * front of it.
 *
 * This is the mobile twin of the extension's MV3 service worker, and it exists
 * for the same reason: the library, progress, dedupe and migration logic is one
 * body of code (`shared/panelflow-core.js`) that Chrome, Android and iOS all
 * execute unchanged. Reimplementing it in Kotlin would mean a merge behaves one
 * way on the phone and another in the browser — which is the bug this whole
 * design exists to make impossible.
 *
 * Keeping it in a WebView also keeps the store (localStorage) on the same side
 * of the native bridge as the code that mutates it. A Room-backed store would
 * turn every merge into a dozen async hops across JNI.
 *
 * Everything the app sends flows through [post]: native answers the handful of
 * messages only native can, and relays the rest here.
 */
object WorkerHost {

    /** Anything that can be sent a reply or an unprompted event. */
    interface Client {
        /** Completes one pending request this client made. `body` is raw JSON. */
        fun deliver(requestId: Long, bodyJson: String)

        /** Unprompted push: `ready`, `notify`, `resumed`. */
        fun emit(event: String, payloadJson: String)
    }

    private val main = Handler(Looper.getMainLooper())
    private val ids = AtomicLong(1)

    /** Requests forwarded to the worker, waiting on its reply. */
    private val inFlight = ConcurrentHashMap<Long, Pair<Client, Long>>()

    /** Replies native itself is waiting on (asking a page for its state). */
    private val nativeWaiters = ConcurrentHashMap<Long, (Any?) -> Unit>()

    /** Shell clients that want `ready`/`notify`; browser pages do not. */
    private val listeners = mutableSetOf<Client>()

    private var worker: WebView? = null
    private var loaded = false

    /** Messages that arrived before the worker finished loading. */
    private val queued = ArrayDeque<Pair<Long, String>>()

    private lateinit var appContext: Context

    // --- lifecycle -----------------------------------------------------------

    @SuppressLint("SetJavaScriptEnabled")
    fun start(context: Context, backendUrl: String) {
        if (worker != null) return
        appContext = context.applicationContext
        val loader = AssetHost.loader(appContext)
        val view = WebView(appContext)
        view.settings.javaScriptEnabled = true
        view.settings.domStorageEnabled = true
        // No user-visible content ever loads here, so nothing else is needed —
        // and every capability left off is one the store cannot be reached by.
        view.settings.allowFileAccess = false
        view.settings.allowContentAccess = false
        view.settings.mediaPlaybackRequiresUserGesture = true

        view.addJavascriptInterface(NativeBridge(WorkerClient), "PanelFlowNative")
        view.webViewClient = object : WebViewClient() {
            override fun shouldInterceptRequest(
                v: WebView,
                request: WebResourceRequest,
            ): WebResourceResponse? = AssetHost.intercept(loader, request)
        }
        worker = view
        view.loadUrl(AssetHost.workerUrl(backendUrl))
    }

    fun addListener(client: Client) = synchronized(listeners) { listeners.add(client) }
    fun removeListener(client: Client) = synchronized(listeners) { listeners.remove(client) }

    /** Settle the local store against the account — called once per cold start. */
    fun boot() = eval("window.PanelFlowWorker.boot()")

    /** The shell asks for this on resume: the library may have changed elsewhere. */
    fun resync() {
        send(object : Client {
            override fun deliver(requestId: Long, bodyJson: String) = broadcast("resumed", "{}")
            override fun emit(event: String, payloadJson: String) = Unit
        }, 0, JSONObject(mapOf("type" to "pullNow")))
    }

    // --- inbound -------------------------------------------------------------

    /**
     * One raw envelope from a WebView. Three shapes arrive here:
     * `{id, msg}` a request, `{reply:{id, body}}` an answer to something we
     * asked for, and `{event, …}` an unprompted push from the worker.
     */
    fun post(client: Client, json: String) {
        val envelope = try { JSONObject(json) } catch (e: Exception) { return }

        // Both the worker answering a forwarded request and a browser page
        // answering a `dispatch` come back as a reply, and both draw their id
        // from the same counter — so one id space, two destinations.
        envelope.optJSONObject("reply")?.let { reply ->
            val id = reply.optLong("id")
            nativeWaiters.remove(id)?.invoke(reply.opt("body")) ?: onWorkerReply(id, reply.opt("body"))
            return
        }
        envelope.optString("event").takeIf { it.isNotEmpty() }?.let { event ->
            onWorkerEvent(event, envelope)
            return
        }
        val msg = envelope.optJSONObject("msg") ?: return
        send(client, envelope.optLong("id"), msg)
    }

    /** Route one message: native handles what only native can, worker gets the rest. */
    private fun send(client: Client, requestId: Long, msg: JSONObject) {
        val handled = NativeMessages.handle(appContext, msg)
        if (handled != null) {
            client.deliver(requestId, handled.toString())
            return
        }
        val workerId = ids.getAndIncrement()
        inFlight[workerId] = client to requestId
        forward(workerId, msg.toString())
    }

    private fun forward(workerId: Long, msgJson: String) = main.post {
        if (!loaded) { queued.addLast(workerId to msgJson); return@post }
        eval("window.PanelFlowWorker.handle($workerId, ${quote(msgJson)})")
    }

    private fun onWorkerEvent(event: String, envelope: JSONObject) = main.post {
        when (event) {
            "loaded" -> {
                loaded = true
                while (queued.isNotEmpty()) {
                    val (id, json) = queued.removeFirst()
                    eval("window.PanelFlowWorker.handle($id, ${quote(json)})")
                }
                boot()
            }
            "ready" -> broadcast("ready", "{}")
            "notify" -> {
                val n = envelope.optJSONObject("notification") ?: JSONObject()
                Notifications.chapter(appContext, n)
                broadcast("notify", n.toString())
            }
            else -> broadcast(event, envelope.toString())
        }
    }

    /** The worker replying: hand the body to whoever originally asked. */
    private object WorkerClient : Client {
        override fun deliver(requestId: Long, bodyJson: String) = Unit
        override fun emit(event: String, payloadJson: String) = Unit
    }

    private fun broadcast(event: String, payloadJson: String) {
        val snapshot = synchronized(listeners) { listeners.toList() }
        for (c in snapshot) c.emit(event, payloadJson)
    }

    // --- native → a page -----------------------------------------------------

    /**
     * Ask an in-app browser page something the extension would ask the active
     * tab (`readerState`, `toggleReader`, `getSeriesMeta`) and await its answer.
     * `onReply` gets null if the page never answers.
     */
    fun ask(page: WebView, msg: JSONObject, timeoutMs: Long = 5_000, onReply: (Any?) -> Unit) {
        val id = ids.getAndIncrement()
        val done = java.util.concurrent.atomic.AtomicBoolean(false)
        nativeWaiters[id] = { body -> if (done.compareAndSet(false, true)) onReply(body) }
        main.post {
            page.evaluateJavascript(
                "window.PanelFlowPage && window.PanelFlowPage.dispatch(${quote(msg.toString())}, $id)",
                null,
            )
        }
        main.postDelayed({
            nativeWaiters.remove(id)
            if (done.compareAndSet(false, true)) onReply(null)
        }, timeoutMs)
    }

    // --- worker replies ------------------------------------------------------

    /** Called by [NativeBridge] when the worker answers a forwarded request. */
    internal fun onWorkerReply(workerId: Long, body: Any?) {
        val (client, requestId) = inFlight.remove(workerId) ?: return
        client.deliver(requestId, body?.toString() ?: "null")
    }

    private fun eval(js: String) = main.post { worker?.evaluateJavascript(js, null) }

    /** JSON-quote a string for embedding in a JS expression. */
    internal fun quote(s: String): String = JSONObject.quote(s)
}

File: ios/Sources/WorkerHost.swift
import WebKit

/// The offscreen `WKWebView` that runs the shared store core, and the
/// switchboard in front of it.
///
/// This is the iOS twin of the Chrome extension's MV3 service worker and of
/// Android's `WorkerHost`, and all three exist for the same reason: the
/// library, progress, dedupe and migration logic is one body of code
/// (`shared/panelflow-core.js`) that every platform executes unchanged.
/// Reimplementing it in Swift would mean a merge behaves one way on the phone
/// and another in the browser — the exact failure this design prevents.
///
/// Keeping it in a web view also keeps the store (`localStorage`) on the same
/// side of the bridge as the code that mutates it. A Core Data store would turn
/// every merge into a dozen async hops across the message handler.
///
/// The view is never added to a window. WebKit throttles timers in views that
/// are off-screen but it does not stop them, and nothing here is timer-driven:
/// every action is a message that arrives and is answered.
/// Anything that can be sent a reply or an unprompted event.
///
/// Top level rather than nested in `WorkerHost` so it names the same thing
/// Android's `WorkerHost.Client` does without depending on a Swift version that
/// allows nested protocols.
@MainActor
protocol PanelFlowClient: AnyObject {
    /// Completes one pending request this client made. `bodyJSON` is raw JSON.
    func deliver(requestID: Int, bodyJSON: String)
    /// Unprompted push: `ready`, `notify`, `resumed`.
    func emit(event: String, payloadJSON: String)
}

@MainActor
final class WorkerHost: NSObject {

    static let shared = WorkerHost()

    typealias Client = PanelFlowClient

    private var worker: WKWebView?
    private var loaded = false
    private var nextID = 1

    /// Requests forwarded to the worker, waiting on its reply.
    private var inFlight: [Int: (client: Client, requestID: Int)] = [:]
    /// Replies native itself is waiting on — asking a page for its state.
    private var nativeWaiters: [Int: (Any?) -> Void] = [:]
    /// Shell clients that want `ready`/`notify`; browser pages do not.
    private var listeners: [WeakClient] = []
    /// Messages that arrived before the worker finished loading.
    private var queued: [(Int, String)] = []

    private struct WeakClient { weak var value: Client? }

    // MARK: - lifecycle

    func start(backendURL: String) {
        guard worker == nil else { return }

        let config = WKWebViewConfiguration()
        let controller = WKUserContentController()
        controller.add(MessageProxy(client: WorkerReplyClient()), name: "panelflow")
        config.userContentController = controller
        // The store lives here, so it must be the persistent store and not a
        // per-launch one — a non-persistent data store would hand the user an
        // empty library after every cold start.
        config.websiteDataStore = .default()

        let view = WKWebView(frame: .zero, configuration: config)
        worker = view

        guard let base = Bundle.main.url(forResource: "worker", withExtension: "html",
                                         subdirectory: "www")
        else { return }
        // The backend is passed on the URL rather than baked into worker.js:
        // both platforms set it the same way, and it stays a *default* that the
        // user's own stored setting overrides.
        var components = URLComponents(url: base, resolvingAgainstBaseURL: false)
        components?.queryItems = [URLQueryItem(name: "backend", value: backendURL)]
        let target = components?.url ?? base
        view.loadFileURL(target, allowingReadAccessTo: base.deletingLastPathComponent())
    }

    func addListener(_ client: Client) {
        listeners.removeAll { $0.value == nil || $0.value === client }
        listeners.append(WeakClient(value: client))
    }

    func removeListener(_ client: Client) {
        listeners.removeAll { $0.value == nil || $0.value === client }
    }

    /// The shell asks for this on foreground: the library may have moved on.
    func resync() {
        send(client: ResyncClient(host: self), requestID: 0, msg: ["type": "pullNow"])
    }

    // MARK: - inbound

    /// One raw envelope from a web view. Three shapes arrive here: `{id, msg}`
    /// a request, `{reply: {id, body}}` an answer to something we asked for,
    /// and `{event, …}` an unprompted push from the worker.
    func post(client: Client, json: String) {
        guard let data = json.data(using: .utf8),
              let envelope = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        else { return }

        // Both the worker answering a forwarded request and a browser page
        // answering a `dispatch` come back as a reply, and both draw their id
        // from the same counter — one id space, two destinations.
        if let reply = envelope["reply"] as? [String: Any], let id = reply["id"] as? Int {
            if let waiter = nativeWaiters.removeValue(forKey: id) {
                waiter(reply["body"])
            } else {
                onWorkerReply(id, body: reply["body"])
            }
            return
        }
        if let event = envelope["event"] as? String {
            onWorkerEvent(event, envelope: envelope)
            return
        }
        guard let msg = envelope["msg"] as? [String: Any] else { return }
        send(client: client, requestID: envelope["id"] as? Int ?? 0, msg: msg)
    }

    /// Route one message: native handles what only native can, worker gets the rest.
    private func send(client: Client, requestID: Int, msg: [String: Any]) {
        if let handled = NativeMessages.handle(msg) {
            client.deliver(requestID: requestID, bodyJSON: encode(handled))
            return
        }
        let workerID = nextID; nextID += 1
        inFlight[workerID] = (client, requestID)
        forward(workerID, msgJSON: encode(msg))
    }

    private func forward(_ workerID: Int, msgJSON: String) {
        guard loaded else { queued.append((workerID, msgJSON)); return }
        eval("window.PanelFlowWorker.handle(\(workerID), \(quote(msgJSON)))")
    }

    private func onWorkerEvent(_ event: String, envelope: [String: Any]) {
        switch event {
        case "loaded":
            loaded = true
            let pending = queued; queued = []
            for (id, json) in pending {
                eval("window.PanelFlowWorker.handle(\(id), \(quote(json)))")
            }
            eval("window.PanelFlowWorker.boot()")
        case "ready":
            broadcast("ready", payloadJSON: "{}")
        case "notify":
            let n = envelope["notification"] as? [String: Any] ?? [:]
            Notifications.chapter(n)
            broadcast("notify", payloadJSON: encode(n))
        default:
            broadcast(event, payloadJSON: encode(envelope))
        }
    }

    private func onWorkerReply(_ workerID: Int, body: Any?) {
        guard let pending = inFlight.removeValue(forKey: workerID) else { return }
        pending.client.deliver(requestID: pending.requestID,
                               bodyJSON: body.map(encodeAny) ?? "null")
    }

    private func broadcast(_ event: String, payloadJSON: String) {
        listeners.removeAll { $0.value == nil }
        for entry in listeners { entry.value?.emit(event: event, payloadJSON: payloadJSON) }
    }

    // MARK: - native → a page

    /// Ask an in-app browser page something the extension would ask the active
    /// tab (`readerState`, `toggleReader`, `getSeriesMeta`) and await its
    /// answer. `onReply` gets nil if the page never answers.
    func ask(page: WKWebView, msg: [String: Any], timeout: TimeInterval = 5,
             onReply: @escaping (Any?) -> Void) {
        let id = nextID; nextID += 1
        var answered = false
        nativeWaiters[id] = { body in
            guard !answered else { return }
            answered = true
            onReply(body)
        }
        page.evaluateJavaScript(
            "window.PanelFlowPage && window.PanelFlowPage.dispatch(\(quote(encode(msg))), \(id))"
        )
        DispatchQueue.main.asyncAfter(deadline: .now() + timeout) { [weak self] in
            self?.nativeWaiters.removeValue(forKey: id)
            guard !answered else { return }
            answered = true
            onReply(nil)
        }
    }

    // MARK: - helpers

    private func eval(_ js: String) { worker?.evaluateJavaScript(js) }

    /// The worker's own replies land here; the routing already happened in `post`.
    private final class WorkerReplyClient: Client {
        func deliver(requestID: Int, bodyJSON: String) {}
        func emit(event: String, payloadJSON: String) {}
    }

    private final class ResyncClient: Client {
        private weak var host: WorkerHost?
        init(host: WorkerHost) { self.host = host }
        func deliver(requestID: Int, bodyJSON: String) {
            MainActor.assumeIsolated { host?.broadcast("resumed", payloadJSON: "{}") }
        }
        func emit(event: String, payloadJSON: String) {}
    }
}

// MARK: - JSON helpers

func encode(_ value: [String: Any]) -> String {
    guard let data = try? JSONSerialization.data(withJSONObject: value),
          let s = String(data: data, encoding: .utf8) else { return "{}" }
    return s
}

func encodeAny(_ value: Any) -> String {
    if JSONSerialization.isValidJSONObject(value),
       let data = try? JSONSerialization.data(withJSONObject: value),
       let s = String(data: data, encoding: .utf8) { return s }
    return "null"
}

/// JSON-quote a string for embedding in a JS expression.
func quote(_ s: String) -> String {
    guard let data = try? JSONSerialization.data(withJSONObject: [s]),
          let wrapped = String(data: data, encoding: .utf8) else { return "\"\"" }
    return String(wrapped.dropFirst().dropLast())
}
